lang/values.md

```diff
@@ -466,9 +466,9 @@ impl<M: Memory> TypedMemory for M {
             (Value::Ptr(ptr), Type::Ptr(ptr_type)) => Value::Ptr(self.retag_ptr(ptr, ptr_type, fn_entry)?),
             // recurse into tuples/arrays/enums
             (Value::Tuple(vals), Type::Tuple { fields, .. }) =>
-                Value::Tuple(vals.iter().zip(fields).map(|(val, (_offset, ty))| self.retag_val(val, ty, fn_entry)).try_collect()?),
```

Comment:
I added this `List::zip`.
Generally I think all those `.iter()` calls make little sense for a spec.

Reply:
I'm happy to open another PR to get rid of all unnecessary `.iter()` calls, by implementing some `List` methods.
You agree?

Reply:
What is the type of all these functions?

Of course we could avoid iterators entirely, and have a bunch of functions that just create lists. But if that means creating more intermediate lists, that sounds bad for the interpreter -- we don't want `list1.zip(list2).try_map(...)` to construct an intermediate list, I think.

So what are the operations we need?
- `fn try_map(List<T>, impl Fn(T) -> Result<U, E>) -> Result<List<U>, E>`
- `fn try_zip_map(List<T1>, List<T2>, impl Fn(T1, T2) -> Result<U, E>) -> Result<List<U>, E>`

anything else?

Reply:
> we don't want list1.zip(list2).try_map(...) to construct an intermediate list, I think.

Interesting, I expected you'd prefer the conciseness over the performance benefit from using iterators in that regard.

Then I guess doing `List::zip(impl IntoIterator) -> impl Iterator` and implementing `IteratorExt::try_map` for an extension trait for `Iterator`s would be the best of both worlds?

Then the code stays `list1.zip(list2).try_map` but it's quick.

Reply:
For now I reverted my change here. We'll do those things in another PR.

Reply:
> For now I reverted my change here. We'll do those things in another PR.

Okay.

If `try_zip_map` is the only thing we will need to completely remove iterators from MiniRust, it might be worth it, despite being rather ad-hoc.